Add support to track events

diegoserranoa opened the pull request:
Add support for Segment's track method. This is to support a new feature
by Wootric for event-based surveying.

### Changes
- Add track() method
- Refactor identify() method
- Add tests

**PS:** I can give more context but I don't know if you guys have any special
format when creating PRs :)

diegoserranoa committed:
Add support to track events

Add support for Segment's track method. This is to support a new feature
by Wootric for event-based surveying.

Changes
- Add track() method
- Refactor identify() method
- Add tests

PS: I can give more context but I don't know if you guys have any special
format when creating PRs :)

diegoserranoa committed:
Add support to track events

Add support for Segment's track method. This is to support a new feature
by Wootric for event-based surveying.

Changes
- Add track() method
- Refactor identify() method
- Add tests

PS: I can give more context but I don't know if you guys have any special
format when creating PRs :)

diegoserranoa commented:
Tests are passing locally, I tried rebuilding in CI but I don't have permissions :)

prabhatjha commented:
Hello -- are you able to give us an ETA on when you will be able to take a look at it? Thanks.

diegoserranoa commented:
Hi guys @ccnixon or @hankim813! Didn't know which one to mention in this PR hehe can we get a little help here? We've introduced a new feature that is needed by a customer!

About the tests, like I mentioned ☝️ they're all passing in my local machine but don't know why they're failing in CI.

Thanks for the help!

PS: for future PRs who should we contact for reviews?

emiliogomezlavin commented:
Adding @segment-integrations/destinations for review

ccnixon commented:
Thanks @diegoserranoa! let us take a look at this tomorrow.

prabhatjha commented:
@ccnixon Just a friendly reminder to have you or team's 👀 on this PR. 🙏

ccnixon commented on lib/index.js:
```diff
@@ -16,6 +17,7 @@ var omit = require('omit');
 var Wootric = module.exports = integration('Wootric')
   .assumesPageview()
   .option('accountToken', '')
+  .global('event_name')
```
Hey wootric team - is `event_name` a global variable set by the sdk?

diegoserranoa replied:
`event_name` is set by `track()`. So I'm guessing this doesn't go here? hehe

ccnixon commented on lib/index.js:
```diff
@@ -8,6 +8,7 @@ var foldl = require('@ndhoule/foldl');
 var is = require('is');
 var integration = require('@segment/analytics.js-integration');
 var omit = require('omit');
+var newDate = require('new-date');
```
We would like to avoid adding dependencies whenever possible. Is there a reason we cant use the native `Date` constructor?

diegoserranoa replied:
I'm using this to make `track()` work as similar as `trait()`. In `trait()`, we have a `createdAt` var that gets its value from `identify.created()`. Since the `track` object doesn't have a `created()` (at least not that why I know of from logging the object and searching the docs), we need to parse the date ourselves from the properties. I don't have the code right now, but if I'm not wrong, I remember `created()` used `newDate` to convert date, which I like since `newDate` allows extra input types. :)

tysonmote replied:
@diegoserranoa The track object should have a `createdAt` timestamp. Does that seem like it will work for you?

diegoserranoa commented:
@ccnixon how do I get the tests to pass? Are they failing because of the `new-date` dependency?

tysonmote commented:
Hi all -- sorry for the delay here! We'll pick this one back up this week!

@diegoserranoa In the meantime, do you mind pasting the test failure you're getting here?

diegoserranoa committed:
Remove dependency and change created_at

diegoserranoa committed:
Remove dependency and change created_at

diegoserranoa commented:
Hi @tysonmote. I pushed another commit: I removed the `newDate` dependency and I'm using now the timestamp that the track object has. Tests keep failing though, most of the errors have this description:

`The environment you requested was unavailable`

https://circleci.com/gh/segment-integrations/analytics.js-integration-wootric/329?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Peripheral1994 commented:
Hey @diegoserranoa - It looks like some old creds were left in the circle.yml file. Can you try deleting those two lines on your branch and retrying? Sorry for this test issue, it looks like those are overriding the actual creds we have in Circle.

Peripheral1994 commented:
Specifically `SAUCE_ACCESS_KEY` and `SAUCE_USERNAME`

diegoserranoa commented:
Sure @Peripheral1994! I'll try that. Thanks for the help!

diegoserranoa committed:
Delete old circle.yml creds

Peripheral1994 commented:
Thanks for helping us get this back in shape! I'm going to make a couple config changes in master here that will hopefully get us back on the right track to deploying this ASAP, one moment.

diegoserranoa commented:
Thanks @Peripheral1994!

Peripheral1994 committed:
Update CircleCI Creds

The old creds were failing - these are good for the interim until a more permanent solution can be reached.

Peripheral1994 commented:
@diegoserranoa - Try merging in the changes from #14 - that should get those tests running!

diegoserranoa committed:
Merge branch 'support-track-method' of https://github.com/diegoserranoa/analytics.js-integration-wootric into support-track-method

diegoserranoa commented:
Test passed @Peripheral1994! 🙏 🙌

Peripheral1994 commented:
@diegoserranoa - Perfect! One last question, I see a lot of "LOG: 'Wootric: Warning: Remove survey immediately before deploying or else customers will be surveyed repeatedly.'" - Are these concerning, or just purely artifacts of the test environment?

diegoserranoa commented:
@Peripheral1994 - This is just a warning log and it's showing up because that var is being set to `true` before each test https://github.com/diegoserranoa/analytics.js-integration-wootric/blob/18a18af6bdb40fd2329f44428b2270cfdb797530/test/index.test.js#L23

Peripheral1994 commented:
I dig it! I'll see if I can't get this out tonight, else first thing in the morning. :)

diegoserranoa commented:
Thanks a lot @Peripheral1994! Let us know if you need more help from our side.

prabhatjha commented:
@Peripheral1994 Ready to 🚢 🚀 ? ;-)

Peripheral1994 commented:
Hey @prabhatjha - I just got into the office, this is first in my queue to send out. :)

jrrera commented:
Hi everyone, this implementation of the `track` method appears to overwrite the existing properties that may have been included from the `identify` method to the `wootricSettings.properties` object. Wouldn't this cause issues for Segment users who expect properties included in `identify` to be sent to Wootric who also use the `track` method before the survey is submitted? In other words:

identify w/ properties --> track w/ properties --> wootric survey pops up and is submitted
^ i would expect this flow to include properties from identify given the Segment/Wootric documentation, but they are overwritten by the track call.

diegoserranoa commented:
Hello @tysonmote @ccnixon @Peripheral1994
While checking the code again, I realized that using the `timestamp` that comes with the track object is not exactly what we need. I will create another PR soon to make it work like I implemented it before, getting the `created_at` from the properties. I may also need to add the dependency that I was using. 

I will then need your help for another review 🙏
